// @logicflow/extension/cjs/bpmn/constant.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.theme = exports.UserTaskConfig = exports.ServiceTaskConfig = exports.ExclusiveGatewayConfig = exports.EndEventConfig = exports.StartEventConfig = void 0;
exports.StartEventConfig = {
    width: 40,
    height: 40,
};
exports.EndEventConfig = {
    width: 40,
    height: 40,
};
exports.ExclusiveGatewayConfig = {
    width: 40,
    height: 40,
};
exports.ServiceTaskConfig = {
    width: 100,
    height: 80,
};
exports.UserTaskConfig = {
    width: 100,
    height: 80,
};
exports.theme = {
    rect: {
        radius: 5,
        stroke: 'rgb(24, 125, 255)',
    },
    circle: {
        r: 18,
        stroke: 'rgb(24, 125, 255)',
    },
    polygon: {
        stroke: 'rgb(24, 125, 255)',
    },
    polyline: {
        stroke: 'rgb(24, 125, 255)',
        hoverStroke: 'rgb(24, 125, 255)',
        selectedStroke: 'rgb(24, 125, 255)',
    },
    edgeText: {
        background: {
            fill: 'white',
            height: 14,
            stroke: 'transparent',
            radius: 3,
        },
    },
};
